src/rust/engine/options/src/lib.rs

```diff
@@ -200,6 +200,30 @@ pub enum Source {
     Flag,
 }
 
+// NB: Must mirror the Rank enum in src/python/pants/option/ranked_value.py.
+pub enum Rank {
+    _NONE = 0,          // Unused, exists for historical Python compatibility reasons.
+    HARDCODED = 1,      // The default provided at option registration.
+    _CONFIGDEFAULT = 2, // Unused, exists for historical Python compatibility reasons.
```

Comment:
Not directly related to this PR, but... is it worth adding `Source::ConfigDefault` (and thus make this used), to model the `[DEFAULT]` section more explicitly?

Reply:
I considered it. As I recall I decided against because it didn't seem like the right conceptual model (despite what the Python code does, which feels more like a hack). I don't think this is how a user would think of it, as two sources. 

But maybe it's worth doing to get rid of the section merging code?

Reply:
It's something we can revisit later, for sure.

Reply:
Yeah revisiting is fine. Maybe the error messages/help descriptions end clearer if we have the exact sources? 


However, this leads me to think about something else: what's the behaviour if we have two config files both with `[DEFAULT]`?

```toml
# pants.toml
[DEFAULT]
opt = [1]

[foo]
opt.add = [2]

# pants.ci.toml
[DEFAULT]
opt = [3]

[foo]
opt.add = [4]
```

I don't know what the answer _should_ be... Maybe `[3, 2, 4]`?

Most importantly... do the Rust and Python code align?

Reply:
Oh hmm, yes! Now that you mention it, I think this was on the edge of my awareness when I decided against (not so much because of this sort of example, which I didn't think of directly, but because of how it would complicate the derivation metadata).

Currently Rust gives `[3, 4]`, and Python gives `[3, 2, 4]`.

I think the answer should be `[3, 4]`: The DEFAULT section is a legacy of `pants.ini`: in INI files, each file falls back to the DEFAULT section if the requested section has no value at the requested key, *as if it was in that section*. So I'd (strongly) argue that the Python behavior is a bug. In fact, IIRC we used to merge the config files into a single logical Config before passing them to the higher-level OptionsParser, but then CONFIGDEFAULT as a source was introduced (by me) to simplify things. So I would say that is were I introduced this bug.

(Note that we don't document the DEFAULT section as providing values, only as providing interpolation variables, but the INI behavior is what was intended).

In fact, I'd argue that merging across [`DEFAULT`] and `[scope]` is probably not implied by the INI contract - but that gets complicated by our use of `.add` - since now the key is different and so technically INI would see them both, but if instead of `.add` we did `opt = "+[2]"`, it would make sense for that to cause config to ignore the DEFAULT value, since as strings this value shadows the DEFAULT one.

Ooof, what a mess...